source/opt/dead_branch_elim_pass.h

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2017 The Khronos Group Inc.
+// Copyright (c) 2017 Valve Corporation
+// Copyright (c) 2017 LunarG Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_OPT_DEAD_BRANCH_ELIM_PASS_H_
+#define LIBSPIRV_OPT_DEAD_BRANCH_ELIM_PASS_H_
+
+
+#include <algorithm>
+#include <map>
+#include <queue>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+
+#include "basic_block.h"
+#include "def_use_manager.h"
+#include "module.h"
+#include "pass.h"
+
+namespace spvtools {
+namespace opt {
+
+// See optimizer.hpp for documentation.
+class DeadBranchElimPass : public Pass {
+
+  using cbb_ptr = const ir::BasicBlock*;
+
+ public:
+   using GetBlocksFunction =
+     std::function<std::vector<ir::BasicBlock*>*(const ir::BasicBlock*)>;
+
+  DeadBranchElimPass();
+  const char* name() const override { return "dead-branch-elim"; }
+  Status Process(ir::Module*) override;
+
+ private:
+  // Returns the id of the merge block declared by a merge instruction in 
+  // this block, if any.  If none, returns zero.
+  uint32_t MergeBlockIdIfAny(const ir::BasicBlock& blk) const;
+
+  // Compute structured successors for function |func|.
```

Comment:
Ok, here's a concrete example of my concern. (See point 3 in previous comment).

Example fragment shader:
```
#version 450
void main() {
  bool c;
  bool d;
  while(c) {
    if(d) {
     continue;
    }
    if(false) {  // setting this up to be killed.  branches out to an outer continue target.
     continue;
    }
    discard;
  }
}
```
Compiles to: 
```
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %main "main"
               OpExecutionMode %main OriginUpperLeft
               OpSource GLSL 450
               OpName %main "main"
               OpName %c "c"
               OpName %d "d"
       %void = OpTypeVoid
          %3 = OpTypeFunction %void
       %bool = OpTypeBool
%_ptr_Function_bool = OpTypePointer Function %bool
      %false = OpConstantFalse %bool
       %main = OpFunction %void None %3
          %5 = OpLabel
          %c = OpVariable %_ptr_Function_bool Function
          %d = OpVariable %_ptr_Function_bool Function
               OpBranch %6
          %6 = OpLabel
               OpLoopMerge %8 %9 None
               OpBranch %10
         %10 = OpLabel
         %14 = OpLoad %bool %c
               OpBranchConditional %14 %7 %8
          %7 = OpLabel
         %16 = OpLoad %bool %d
               OpSelectionMerge %18 None
               OpBranchConditional %16 %17 %18
         %17 = OpLabel
               OpBranch %9
         %18 = OpLabel
               OpSelectionMerge %22 None
               OpBranchConditional %false %21 %22
         %21 = OpLabel
               OpBranch %9
         %22 = OpLabel
               OpKill
          %9 = OpLabel
               OpBranch %6
          %8 = OpLabel
               OpReturn
               OpFunctionEnd
```
Control flow graph is:
![stray](https://user-images.githubusercontent.com/13054574/27839661-047dc1ca-60c1-11e7-8b16-783f42dcb292.png)


Then running through this optimization pass, I get:
```
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %main "main"
               OpExecutionMode %main OriginUpperLeft
               OpSource GLSL 450
               OpName %main "main"
               OpName %c "c"
               OpName %d "d"
       %void = OpTypeVoid
          %3 = OpTypeFunction %void
       %bool = OpTypeBool
%_ptr_Function_bool = OpTypePointer Function %bool
      %false = OpConstantFalse %bool
       %main = OpFunction %void None %3
          %5 = OpLabel
          %c = OpVariable %_ptr_Function_bool Function
          %d = OpVariable %_ptr_Function_bool Function
               OpBranch %6
          %6 = OpLabel
               OpLoopMerge %8 %9 None
               OpBranch %10
         %10 = OpLabel
         %14 = OpLoad %bool %c
               OpBranchConditional %14 %7 %8
          %7 = OpLabel
         %16 = OpLoad %bool %d
               OpSelectionMerge %18 None
               OpBranchConditional %16 %17 %18
         %17 = OpLabel
               OpBranch %9
         %18 = OpLabel
               OpBranch %22
         %22 = OpLabel
               OpKill
          %8 = OpLabel
               OpReturn
               OpFunctionEnd
```
That's invalid because block %9 has been swept away but is mentioned as the continue target of the loop, and as the branch target of block %17.

The problem is that you determine that block %21 is dead (that's right) and all the blocks it dominates should be swept away.  But your algorithm incorrectly determines that block %9 should be included in that set.